go/arrow/decimal256/decimal256.go

```diff
@@ -0,0 +1,569 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package decimal256
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"math/big"
+
+	"github.com/apache/arrow/go/v10/arrow/decimal128"
+	"github.com/apache/arrow/go/v10/arrow/internal/debug"
+)
+
+const (
+	MaxPrecision = 76
+	MaxScale     = 76
+)
+
+type Num struct {
+	// arr[0] is the lowest bits, arr[3] is the highest bits
+	arr [4]uint64
+}
+
+// New returns a new signed 256-bit integer value where x1 contains
+// the highest bits with the rest of the values in order down to the
+// lowest bits
+//
+//   ie: New(1, 2, 3, 4) returns with the elements in little-endian order
+//       {4, 3, 2, 1} but each value is still represented as the native endianness
+func New(x1, x2, x3, x4 uint64) Num {
+	return Num{[4]uint64{x4, x3, x2, x1}}
+}
+
+func (n Num) Array() [4]uint64 { return n.arr }
+
+func (n Num) LowBits() uint64 { return n.arr[0] }
+
+func FromDecimal128(n decimal128.Num) Num {
+	var topBits uint64
+	if n.Sign() < 0 {
+		topBits = math.MaxUint64
+	}
+	return New(topBits, topBits, uint64(n.HighBits()), n.LowBits())
+}
+
+func FromU64(v uint64) Num {
+	return Num{[4]uint64{v, 0, 0, 0}}
+}
+
+func FromI64(v int64) Num {
+	switch {
+	case v > 0:
+		return New(0, 0, 0, uint64(v))
+	case v < 0:
+		return New(math.MaxUint64, math.MaxUint64, math.MaxUint64, uint64(v))
+	default:
+		return Num{}
+	}
+}
+
+func (n Num) Negate() Num {
+	var carry uint64 = 1
+	for i := range n.arr {
+		n.arr[i] = ^n.arr[i] + carry
+		if n.arr[i] != 0 {
+			carry = 0
+		}
+	}
+	return n
+}
+
+func FromFloat32(v float32, prec, scale int32) (Num, error) {
+	debug.Assert(prec > 0 && prec <= 76, "invalid precision for converting to decimal256")
+
+	if math.IsInf(float64(v), 0) {
+		return Num{}, fmt.Errorf("cannot convert %f to decimal256", v)
+	}
+
+	if v < 0 {
+		dec, err := fromPositiveFloat32(-v, prec, scale)
+		if err != nil {
+			return dec, err
+		}
+		return dec.Negate(), nil
+	}
+	return fromPositiveFloat32(v, prec, scale)
+}
+
+func fromPositiveFloat32(v float32, prec, scale int32) (Num, error) {
+	var pscale float32
+	if scale >= -76 && scale <= 76 {
+		pscale = float32PowersOfTen[scale+76]
```

Comment:
Wouldn't it be better to cast `v` to float64 and then call `fromPositive64`? You would 1) remove a lot of code 2) get better precision (`float32PowersOfTen` contains some zeros and infinities for very small or large scales)

Reply:
So i tried that originally when I was writing the `ToFloat` functionalities for decimal128 and found (*very*) small precision issues which I assumed would carry through to the `From` functions. though right now trying this locally I can't reproduce those precision issues, maybe it was a windows thing? I don't know. That said, given that those precision issues would be outside the bounds of what is considered even reasonable for a float, I agree with you that it's probably just easier/better to cast to float64 and use the `fromPositiveFloat64` / `tofloat64Positive` functions and remove the excess code. I'll go do that.